### PR TITLE
#541 (4) - Docs

### PR DESCRIPTION
**Pull request by treshugart**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -378,31 +404,6 @@ The parameters passed to the function are:
 
 
 
-##### `created`
```
Comment (treshugart) [inline README.md]: No longer exists.

File: README.md [hunk 2]
```diff
@@ -595,76 +619,29 @@ When the property is initialised, `oldValue` will always be `undefined` and `new
...
+    skate.vdom.p(`My name is ${elem.tagName}.`);
   }
 });
 ```
 
 The only argument passed to `render` is the component element. In this case that is `<my-component>`.
 
-
-
-##### Custom renderers
```
Comment (treshugart) [inline README.md]: Not supported because now we're being opinionated.

File: README.md [hunk 3]
```diff
@@ -739,98 +716,13 @@ Skate will automatically detect the native prototype for the element that you ar
 
 
 
-#### `type`
```
Comment (treshugart) [inline README.md]: No longer supported; custom elements only.

File: README.md [hunk 4]
```diff
@@ -1077,12 +973,79 @@ skate.init(element1, element2);
 
 
 
+### `link (element, propSpec)`
```
Comment (treshugart) [inline README.md]: New function.

File: README.md [hunk 5]
```diff
@@ -1232,26 +1195,175 @@ This does not solve the situation where you want to be notified of future elemen
 
 
 
-### `render (element)`
+### `state (elem[, state])`
```
Comment (treshugart) [inline README.md]: New function.

File: README.md [hunk 6]
```diff
@@ -1232,26 +1195,175 @@ This does not solve the situation where you want to be notified of future elemen
...
+});
+
+// Only returns props you've defined on your component.
+// { prop1: 'value 1' }
+skate.state(elem);
+```
+
+
+
+### `vdom`
```
Comment (treshugart) [inline README.md]: New functions.

File: README.md [hunk 7]
```diff
@@ -1335,7 +1447,10 @@ This is because Mutation Observers queue a [microtask](https://jakearchibald.com
 
 ## Web Component Differences
 
-Skate implements the [Custom Element spec](http://w3c.github.io/webcomponents/spec/custom/) with a custom API but it does not polyfill the native methods. Since Skate is a custom element library, it does not polyfill [ShadowDOM](http://w3c.github.io/webcomponents/spec/shadow/) or [HTML Imports](http://w3c.github.io/webcomponents/spec/imports/).
+Skate implements the [Custom Element spec](http://w3c.github.io/webcomponents/spec/custom/) with a custom API but it does not polyfill the native methods. Since Skate requires that you BYO your own Shadow DOM polyfill. Some polyfills are:
```
Comment (treshugart) [inline README.md]: Needs revising. Should highlight: 

- it internally polyfills custom elements but not others and reasoning behind it (custom elements can be provided without touching native methods, shadow dom cannot).
- faster than webcomponentsjs custom element polyfill
- about the same size as custom element polyfill
- far more functionality

File: README.md [hunk 8]
```diff
@@ -1354,7 +1469,7 @@ Like the web component differences mentioned above, Skate only polyfills and val
 
 
 
-## Polymer Differences
+## VS Polymer
```
Comment (treshugart) [inline README.md]: Needs revising. Some points to use:

- Polymer has a fundamentally different programming model (functional etc)
- Performance, referencing that Polymer uses the webcomponentsjs polyfills and ours are faster
- Polyfills are outdated (uses v0 API, Skate is up to date)
- Polymer isn't on NPM, only Bower
- Polymer uses HTML imports for their codebase (list downsides as well as it's probably not going to be a spec anymore)

More?

File: README.md [hunk 9]
```diff
@@ -1399,62 +1515,23 @@ Theoretically, Skate could automatically detect this and extend `path` by defaul
 
 
 
-## Polyfills
```
Comment (treshugart) [inline README.md]: Added section at the top under "Dependencies".

File: README.md [hunk 10]
```diff
@@ -1399,62 +1515,23 @@ Theoretically, Skate could automatically detect this and extend `path` by defaul
...
+
+my-element[defined] {
   opacity: 1;
   transition: opacity .3s ease;
 }
 ```
 
 
 
-## Ignoring Elements
```
Comment (treshugart) [inline README.md]: I don't want people to use this. I believe we're the only ones using it and its behaviour is inconsistent as it doesn't work in native.

File: README.md [hunk 11]
```diff
@@ -1399,62 +1515,23 @@ Theoretically, Skate could automatically detect this and extend `path` by defaul
...
-
-```html
-<div data-skate-ignore>
-  <!-- Everything including the container will be ignored. -->
-</div>
-```
-
-
-
-## Multiple Version Support
```
Comment (treshugart) [inline README.md]: Listed as a bullet point at the top.

File: README.md [hunk 12]
```diff
@@ -1518,81 +1609,6 @@ skate('my-component', {
 
 
 
-### Content Projection
```
Comment (treshugart) [inline README.md]: No longer needed because we're now opinionated.

File: README.md [hunk 13]
```diff
@@ -1518,81 +1609,6 @@ skate('my-component', {
...
-    <my-option>1</my-option>
-    <my-option>2</my-option>
-    <my-option>3</my-option>
-  </div>
-</my-select>
-```
-
-
-
 ### React Integration
```
Comment (treshugart) [inline README.md]: Should this go higher up?

File: README.md [hunk 14]
```diff
@@ -5,37 +5,58 @@
 
 [![Sauce Test Status](https://saucelabs.com/browser-matrix/skatejs.svg)](https://saucelabs.com/u/skatejs)
 
-*This is the README for the master branch and is probably out of sync with the last stable release. To see the README for the last stable release select it from the list of tags.*
+Skate is a library built on top of the [W3C web component specs](https://github.com/w3c/webcomponents) that enables you to write functional and performant web components with a very small footprint.
+
+- Functional rendering pipeline backed by Google's [Incremental DOM](https://github.com/google/incremental-dom).
+- Inherently cross-framework compatible. For example, it works seamlessly with - and complements - React and other frameworks.
+- It's only 8k min+gz and it will only get smaller as more browsers start supporting web components natively.
+- It's very fast. It's roughly 4x as fast as React for similarly written components.
```
Comment (treshugart) [inline README.md]: Is there a better way to say this but still have the clear reference point?

File: README.md [hunk 15]
```diff
@@ -5,37 +5,58 @@
...
+    skate.vdom.text(`Hello, ${elem.name}`);
   }
 });
 ```
 
 Result
 
 ```html
-<my-element>Hello, World!</my-element>
+<my-element>Hello, Bob!</my-element>
```
Comment (nadiam84) [inline README.md]: <x-hello> maybe?

File: README.md [hunk 16]
```diff
@@ -259,17 +285,17 @@ In native custom elements, you must provide the entire prototype for your custom
...
     someMethod: {
-      value: function () {}
+      value () {}
     }
   })
 });
 ```
 
-If your `prototype` doesn't inherit from the base `HTMLElement`, Skate will automatically do this for you.
+In Skate, however, if your `prototype` doesn't inherit from the base `HTMLElement`, Skate will automatically do this for you so you don't have to worry about it.
```
Comment (nadiam84) [inline README.md]: It feels like a comma should go there. "for you, so you don't have to"

File: src/lifecycle/created.js [hunk 17]
```diff
@@ -67,7 +78,7 @@ export default function (opts) {
     }
 
     if (propertyDefinitions) {
-      propertiesCreated(this, propertyDefinitions);
+      iniitaliseProperties(this, propertyDefinitions);
```
Comment (chrisdarroch) [inline src/lifecycle/created.js]: Spelling mistake?

File: README.md [hunk 18]
```diff
@@ -702,11 +679,11 @@ The only argument passed to `detached` is component element. In this case that i
 
 #### `attribute`
 
-Function that is called whenever an attribute is added, updated or removed. This corresponds to the native `attributeChangedCallback` and is *not* called for attributes that exist on the element before it is upgraded, just like with native custom elements. Generally, you'll probably end up using `properties` that have linked attributes instead of this callback.
+Function that is called whenever an attribute is added, updated or removed. This corresponds to the native `attributeChangedCallback`. Generally, you'll probably end up using `properties` that have linked attributes instead of this callback but there are still use cases where this could come in handy.
```
Comment (nadiam84) [inline README.md]: It feels like a comma should be there: "of this callback, but there are still"

File: README.md (hunk 7 above)
Comment (soneymathew) [inline README.md]: I would consider adding some actual metrics or order of magnitude instead of "faster"

Comment (treshugart) [inline README.md]: Good point. The performance tests show this, though. How could we put this into the bullet point as proof?

File: src/lifecycle/created.js (hunk 17 above)
Comment (treshugart) [inline src/lifecycle/created.js]: Yeah, this is fixed in #542. The PRs all contain the work of the last one, that's why they're numbered. I wanted to break the work up into chunks.

**Comment by treshugart** [comment]
This will pass once #544 is merged into master and master is merged into this.

**Pull request merged**